File: backend/mr/src/main/java/io/featurehub/Application.java
package io.featurehub;

import static io.featurehub.rest.Info.APPLICATION_NAME_PROPERTY;

import io.featurehub.events.CloudEventConfigDiscoveryService;
import io.featurehub.health.MetricsHealthRegistration;
import io.featurehub.jersey.FeatureHubJerseyHost;
import io.featurehub.lifecycle.ApplicationLifecycleManager;
import io.featurehub.lifecycle.LifecycleStatus;
import io.featurehub.lifecycle.TelemetryFeature;
import io.featurehub.mr.ManagementRepositoryFeature;
import io.featurehub.mr.dacha2.Dacha2Feature;
import io.featurehub.mr.events.dacha2.CacheApi;
import io.featurehub.web.security.saml.SamlEnvironmentalFeature;
import jakarta.inject.Singleton;
import org.glassfish.jersey.server.ResourceConfig;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class Application {
  private static final Logger log = LoggerFactory.getLogger(Application.class);

  public static void main(String[] args) {
    System.setProperty("user.timezone", "UTC");
    System.setProperty(APPLICATION_NAME_PROPERTY, "management-repository");
    CloudEventConfigDiscoveryService.Companion.addTags("mr");

    try {
      new Application().run();
    } catch (Exception e) {
      log.error("failed", e);
      // force any running Jersey context's to terminate
      ApplicationLifecycleManager.Companion.updateStatus(LifecycleStatus.TERMINATING);
      System.exit(-1);
    }
  }

  private void run() throws Exception {
    // register our resources, try and tag them as singleton as they are instantiated faster
    ResourceConfig config = new ResourceConfig(
      ManagementRepositoryFeature.class,
      SamlEnvironmentalFeature.class,
      TelemetryFeature.class
      );

    MetricsHealthRegistration.Companion.registerMetrics(config, (resourceConfig, locator, binder) -> {
      if (locator != null) {
        binder.bind(locator.getService(CacheApi.class)).to(CacheApi.class).in(Singleton.class);
      }

      if (resourceConfig != null) {
        resourceConfig.register(Dacha2Feature.class);
      }

      return resourceConfig;
    });

    new FeatureHubJerseyHost(config).start();

    log.info("MR Launched - (HTTP/2 payloads enabled!)");

    Thread.currentThread().join();
  }
}

File: backend/party-server/src/main/kotlin/io.featurehub.party/Application.kt
package io.featurehub.party

import io.featurehub.dacha.api.DachaApiKeyService
import io.featurehub.dacha.api.DachaClientFeature
import io.featurehub.dacha.api.DachaClientServiceRegistry
import io.featurehub.edge.EdgeFeature
import io.featurehub.edge.EdgeResourceFeature
import io.featurehub.events.CloudEventConfigDiscoveryService
import io.featurehub.events.kinesis.KinesisEventFeature
import io.featurehub.events.pubsub.PubsubEventFeature
import io.featurehub.health.MetricsHealthRegistration.Companion.registerMetrics
import io.featurehub.jersey.FeatureHubJerseyHost
import io.featurehub.lifecycle.LifecycleListeners
import io.featurehub.lifecycle.LifecyclePriority
import io.featurehub.lifecycle.LifecycleStarted
import io.featurehub.lifecycle.TelemetryFeature
import io.featurehub.mr.ManagementRepositoryFeature
import io.featurehub.mr.dacha2.Dacha2Feature
import io.featurehub.publish.ChannelConstants
import io.featurehub.publish.NATSFeature
import io.featurehub.rest.CacheControlFilter
import io.featurehub.rest.CorsFilter
import io.featurehub.rest.Info.Companion.APPLICATION_NAME_PROPERTY
import io.features.webhooks.features.WebhookFeature
import jakarta.inject.Inject
import org.glassfish.jersey.server.ResourceConfig
import org.slf4j.Logger
import org.slf4j.LoggerFactory

@LifecyclePriority(priority = 2)
class PartyInjectorStarter @Inject constructor(
  private val dachaServiceRegistry: DachaClientServiceRegistry,
  private val apiKeyService: DachaApiKeyService) : LifecycleStarted {
  override fun started() {
    // make sure Edge talks internally to Dacha for the current cache
    dachaServiceRegistry.registerApiKeyService(ChannelConstants.DEFAULT_CACHE_NAME, apiKeyService )
  }
}

class Application {
  private val log = LoggerFactory.getLogger(io.featurehub.Application::class.java)

  @Throws(Exception::class)
  private fun run() {
    // register our resources, try and tag them as singleton as they are instantiated faster
    val config = ResourceConfig(
      NATSFeature::class.java,
      PubsubEventFeature::class.java,
      KinesisEventFeature::class.java,
      CorsFilter::class.java,
      ManagementRepositoryFeature::class.java,
      EdgeResourceFeature::class.java,
      EdgeFeature::class.java,
      io.featurehub.dacha2.Dacha2Feature::class.java,
      DachaClientFeature::class.java,
      Dacha2Feature::class.java, // MR API for dacha2
      TelemetryFeature::class.java,
      CacheControlFilter::class.java,
    )

    if (WebhookFeature.enabled) {
      config.register(WebhookFeature::class.java)
    }

    LifecycleListeners.starter(PartyInjectorStarter::class.java, config)

    registerMetrics(config)

      // check if we should list on a different port
//    registerMetrics(config) { resourceConfig, locator, binder ->
//      if (resourceConfig != null) {
//        resourceConfig.register(Dacha2Feature::class.java)
//      }
//      resourceConfig
//    }

    FeatureHubJerseyHost(config).start()
    log.info("MR Launched - (HTTP/2 payloads enabled!)")

    Thread.currentThread().join()
  }

  companion object {
    private val log: Logger = LoggerFactory.getLogger(Application::class.java)

    @JvmStatic
    fun main(args: Array<String>) {
      System.setProperty("user.timezone", "UTC")
      System.setProperty("dacha1.enabled", "false")
      System.setProperty("dacha2.enabled", "true")
      System.setProperty(APPLICATION_NAME_PROPERTY, "party-server")

      CloudEventConfigDiscoveryService.addTags("dacha2", "edge", "mr")

      try {
        Application().run()
      } catch (e: Exception) {
        log.error("failed", e)
        System.exit(-1)
      }
    }
  }
}
